@xen-orchestra/fs/src/azure.js

```diff
@@ -0,0 +1,221 @@
+  #container
+  #dir
+  #blobServiceClient
+  #containerClient
+
+  constructor(remote, _opts) {
+    super(remote)
+    const { username, path, password, host, protocol } = parse(remote.url)
+
+    if (host) {
+      info('Connecting to Azurite blob storage emulator...')
+      this.#blobServiceClient = BlobServiceClient.fromConnectionString(
+        `DefaultEndpointsProtocol=${protocol};AccountName=${username};AccountKey=${password};BlobEndpoint=${protocol}://${host}/${username}`
+      )
+    } else {
+      info('Connecting to Azure blob storage...')
+      const credentials = new StorageSharedKeyCredential(username, password)
+      this.#blobServiceClient = new BlobServiceClient(`${protocol}://${username}.blob.core.windows.net`, credentials)
+    }
+
+    const parts = split(path)
+    this.#container = parts.shift()
+    this.#dir = join(...parts)
+    this.#containerClient = this.#blobServiceClient.getContainerClient(this.#container)
+    const WITH_RETRY = ['_copy', '_getSize', '_list', '_outputFile', '_read', '_rename', '_unlink', '_writeFile']
+    WITH_RETRY.forEach(functionName => {
+      if (this[functionName] !== undefined) {
+        this[functionName] = pRetry.wrap(this[functionName], {
+          delays: [100, 200, 500, 1000, 2000],
+          when: err => !['SystemInUse'].includes(err?.code),
+          onRetry(error) {
+            warn('retrying method on fs ', {
+              method: functionName,
+              attemptNumber: this.attemptNumber,
+              delay: this.delay,
+              error,
+              file: this.arguments?.[0],
+            })
+          },
+        })
+      }
+    })
+  }
+
+  get type() {
+    return 'azure'
+  }
+  async _sync() {
+    await this.#containerClient.createIfNotExists()
+    await super._sync()
+  }
+
+  async #makePrefix(path) {
+    return path.endsWith('/') ? path : `${path}/`
+  }
+
+  async #isNotEmptyDir(dir) {
+    const prefix = this.#makePrefix(dir)
+    const iterator = this.#containerClient.listBlobsFlat({ prefix }).byPage({ maxPageSize: 1 })
+
+    const { value } = await iterator.next()
+    return (value?.segment?.blobItems?.length ?? 0) > 0
+  }
+
+  async #streamToBuffer(readableStream, buffer) {
+    return new Promise((resolve, reject) => {
+      let bytesRead = 0
+
+      readableStream.on('data', data => {
+        if (bytesRead + data.length <= buffer.length) {
+          data.copy(buffer, bytesRead)
+          bytesRead += data.length
+        } else {
+          reject(new Error('Buffer size exceeded'))
+        }
+      })
+      readableStream.on('end', () => {
+        resolve(bytesRead)
+      })
+      readableStream.on('error', reject)
+    })
+  }
+
+  async _outputStream(file, data) {
+    if (data.length < MIN_BLOCK_SIZE) {
+      await this._writeFile(file, data)
+      return
+    }
+
+    const blobClient = this.#containerClient.getBlockBlobClient(file)
+    const blockCount = Math.ceil(data.length / MAX_BLOCK_SIZE)
```

Comment:
the stream don't always have a length property. Please reuse the parameters of s3._outuputStream `async _outputStream(path, input, { streamLength, maxStreamLength = streamLength, validator })`